src/fsharp/vs/ServiceFormatting/CodeFormatterImpl.fs

```diff
@@ -0,0 +1,713 @@
+        | SynPat.Ands(pats, _range) ->
+            List.forall validatePattern pats
+        | SynPat.LongIdent(_, _, _, constructorArgs, _, _) -> 
+            validateConstructorArgs constructorArgs
+        | SynPat.Tuple(pats, _range) ->
+            List.forall validatePattern pats
+        | SynPat.Paren(pat, _range) ->
+            validatePattern pat
+        | SynPat.ArrayOrList(_isArray, pats, _range) ->
+            List.forall validatePattern pats
+        | SynPat.Record(identPats, _range) ->
+            List.forall (fun (_, pat) -> validatePattern pat) identPats
+        | SynPat.OptionalVal(_ident, _range) -> true
+        | SynPat.IsInst(_typ, _range) -> true
+        | SynPat.QuoteExpr(expr, _range) ->
+            validateExpr expr
+        | SynPat.DeprecatedCharRange _
+        | SynPat.InstanceMember _
+        | SynPat.StructTuple _ -> true
+        | SynPat.FromParseError _ -> false
+
+    and validateConstructorArgs = function
+        | SynConstructorArgs.Pats pats ->
+            List.forall validatePattern pats
+        | SynConstructorArgs.NamePatPairs(identPats, _range) ->
+            List.forall (snd >> validatePattern) identPats
+
+    match ast with
+    | ParsedInput.SigFile _input ->
+        // There is not much to explore in signature files
+        true
+    | ParsedInput.ImplFile input -> 
+        validateImplFileInput input
+
+let formatWith ast moduleName input config =
+    // Use '\n' as the new line delimiter consistently
+    // It would be easier for F# parser
+    let sourceCode = defaultArg input String.Empty
+    let normalizedSourceCode = String.normalizeNewLine sourceCode
+    let formattedSourceCode =
+        Context.create config normalizedSourceCode 
+        |> genParsedInput { ASTContext.Default with TopLevelModuleName = moduleName } ast
+        |> dump
+        |> if config.StrictMode then id else integrateComments normalizedSourceCode
+
+    // Sometimes F# parser gives a partial AST for incorrect input
+    if input.IsSome && String.IsNullOrWhiteSpace normalizedSourceCode <> String.IsNullOrWhiteSpace formattedSourceCode then
+        raise <| FormatException "Incomplete code fragment which is most likely due to parsing errors or the use of F# constructs newer than supported."
+    else formattedSourceCode
+
+let format config sourceCode filePath ast = formatWith ast (Path.GetFileNameWithoutExtension filePath) (Some sourceCode) config
+
+let formatAST ast fileName sourceCode config =
+    let formattedSourceCode = formatWith ast fileName sourceCode config
+        
+    // When formatting the whole document, an EOL is required
+    if formattedSourceCode.EndsWith(Environment.NewLine) then 
+        formattedSourceCode 
+    else 
+        formattedSourceCode + Environment.NewLine
+
+/// Make a position at (line, col) to denote cursor position
+let makePos line col = mkPos line col
+
+/// Make a range from (startLine, startCol) to (endLine, endCol) to select some text
```

Comment:
We should add notes to eventually use the existing versions of this

Reply:
I've not reviewed the code. A quick look at it left an impression that it needs a serious refactoring.